src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -224,6 +226,23 @@ public Worker(String session, ShardWorkerConfig config) throws ConfigurationExce
     this(session, ServerBuilder.forPort(config.getPort()), config);
   }
 
+  public boolean shutDownWorkerGracefully() {
+    isDeregistered = true;
+    logger.log(Level.INFO, "The current worker is deregistered and should be shutdown gracefully!");
+
+    try {
+      while (!pipeline.isEmpty()) {
+        SECONDS.sleep(10);
+      }
+    } catch (InterruptedException e) {
+      logger.log(Level.SEVERE, "The worker gracefully shutdown is interrupted: " + e.getMessage());
+      return false;
+    }
+    logger.log(Level.INFO, "The worker Pipeline is empty now!");
+    // turn off auto scaling protection here
```

Comment:
do we need to call `disableHostProtection` here?

Reply:
oh, my bad, forget to delete the comments. Yeah, that's also my initial thoughts like worker can disable the protection through a grpc call to admin service. But actually, we don't need that because whoever makes the grpc call to inform worker that it should be draining all work from its pipeline and get ready for shutdown will receive the signal and can make call to disable the protection.